Libwrap features

digetx opened the pull request:
I'm thinking about to make record-replay utilizing libwrap. Let's add basic support of DRM IOCTL's tracing for the start, it should be useful anyway.

digetx committed:
libwrap: non-crashing dlsym overload

I've got some annoying crashes and just applied the first googled solution
that worked.

https://stackoverflow.com/questions/15599026/how-can-i-intercept-dlsym-calls-using-ld-preload

digetx committed:
libwrap: print how much time takes IOCTL

This useful feature allows to perform trivial benchmarking using libwrap.

kusma commented on src/libwrap/drm.h:
```diff
@@ -0,0 +1,947 @@
+ */
+
+#ifndef _DRM_H_
+#define _DRM_H_
```
Do we need to commit this header? Can't we source it from some kernel-headers dependency instead?

kusma replied:
After digging a bit, it looks like the alternative is either to copy this header, or provide some build-system switch to provide a path to the kernel-headers. I'm not sure what is nicer, really.

kusma replied:
But maybe give importing of kernel headers its own commit, so it's a bit easier to review?

digetx replied:
I think having a dependency on libdrm means the need to have extra churn with the BLOB environment. Seems easier to just have a local copy. I don't mind at all if you'd decide to re-do it in other way later.

Sure, I'll factor out the headers importing.

kusma replied:
Yeah, I think that's fine. Seems libdrm doesn't expose the kernel uapi as part of their public API, and digging into a source directory seems more fragile than just importing anyway.

kusma commented on src/libwrap/syscall.c:
```diff
@@ -201,34 +218,46 @@ int ioctl(int fd, unsigned long request, ...)
 	printf("%s(addr=%p, length=%zu)\n", __func__, addr, length);
 
-	ret = orig(addr, length);
+	ret = munmap_orig(addr, length);
```
I suspect that most of the changes here should be separated out of this commit...

digetx replied:
Ok, I'll factor out couple of chunks.

kusma commented on src/libwrap/cdma_parse.c:
```diff
@@ -0,0 +1,312 @@
+	uint32_t *words;
+};
+
+static enum nvhost_opcode nvhost_stream_get_opcode(struct nvhost_stream *stream)
```
This is not a big deal at all, but I think at this point it would make sense to rename these methods to cdma_foo rather than nvhost_foo, as they aren't really nvhost-specific anymore, no...

digetx replied:
ok

kusma reviewed (commented):
Generally very nice :)

I think I have some of the same patches lying around, I'll compare when I get home :)

digetx committed:
libwrap: import DRM UAPI headers

They are needed for the DRM tracing implementation.

digetx committed:
libwrap: factor out original mmap/munmap functions

They are needed for the upcoming support of DRM tracing. In particular it
will be used by DRM tracer for having a local mapping of the BO's, so that
the content of BO's could be accessed at any time by the tracer.

digetx committed:
libwrap: add overload for open64()

libdrm is compiled with the 'LARGEFILE' support by default and so the 'open'
function is substituted with 'open64'. This is needed for DRM tracer to trace
anything that uses libdrm and alike.

digetx committed:
libwrap: support DRM tracing

Add basic support for DRM IOCTL's tracing.

digetx committed:
libwrap: expose registers data from CDMA parser

This will be used by upcoming disassembler.

digetx committed:
libwrap: disassemble shaders from traced streams

Now in addition to registers writes tracing, these writes will be assembled
into a shader program state that will be disassembled on the 'DRAW' invocation
and result printed to stdout. This should be useful for debugging stuff in place
and for RE'ing.

digetx commented:
@kusma I've updated commits + implemented disasm.

kusma reviewed (approved):
Cool :D

kusma commented on src/libwrap/Makefile.am:
```diff
@@ -1,18 +1,43 @@
+
 if ENABLE_RNN
-libgrate_wrap_la_CPPFLAGS = -DENABLE_RNN
+libgrate_wrap_la_CFLAGS += -DENABLE_RNN
```
What's this about? This is a CPP (C pre-processor) flag, and CPPFLAGS seems like the right place for it...

digetx replied:
Yes, it is a CPP flag. Do we have any CPP code? Although, I don't know what's the actual difference if CPP flags are used for C as well. Why should we prefer CPPFLAGS over CFLAGS?

kusma replied:
I wonder if you're confusing CPPFLAGS and CXXFLAGS. CPPFLAGS is for the preprocessor flags. It can (in theory at least) get applied in more places, for instance if we're comiling C++ or non-pre-processed ASM code. It's not a big deal or anything, it just seems strange to me to change this, as CPPFLAGS seems slightly more correct.

digetx replied:
I thought about C++, indeed.

digetx replied:
CPPFLAGS returned.

kusma reviewed (approved):
Nice!
